apollo-runtime/src/main/java/com/apollographql/apollo/internal/batch/BatchHttpCall.kt

```diff
@@ -0,0 +1,130 @@
+package com.apollographql.apollo.internal.batch
+
+import com.apollographql.apollo.api.ScalarTypeAdapters
+import com.apollographql.apollo.api.cache.http.HttpCache
+import com.apollographql.apollo.api.cache.http.HttpCachePolicy
+import com.apollographql.apollo.api.internal.Optional
+import com.apollographql.apollo.api.internal.json.BufferedSourceJsonReader
+import com.apollographql.apollo.api.internal.json.JsonWriter
+import com.apollographql.apollo.api.internal.json.ResponseJsonStreamReader
+import com.apollographql.apollo.api.internal.json.Utils
+import com.apollographql.apollo.api.internal.json.writeArray
+import com.apollographql.apollo.cache.ApolloCacheHeaders
+import com.apollographql.apollo.exception.ApolloException
+import com.apollographql.apollo.interceptor.ApolloInterceptor
+import com.apollographql.apollo.internal.interceptor.ApolloServerInterceptor
+import okhttp3.Call
+import okhttp3.Callback
+import okhttp3.HttpUrl
+import okhttp3.Request
+import okhttp3.RequestBody
+import okhttp3.Response
+import okhttp3.ResponseBody
+import okio.Buffer
+import okio.ByteString
+import java.io.IOException
+import java.nio.charset.Charset
+
+/**
+ * Represents a single HTTP call that requests data for a list of queries at once
+ */
+class BatchHttpCall(
+    private val queryList: List<QueryToBatch>,
+    private val serverUrl: HttpUrl,
+    private val httpCallFactory: Call.Factory,
+    private val scalarTypeAdapters: ScalarTypeAdapters
+) {
+
+  fun execute() {
+    val queryRequestBodyList = mutableListOf<ByteString>()
+
+    for (query in queryList) {
+      // Trigger fetch started callback
+      query.callback.onFetch(ApolloInterceptor.FetchSourceType.NETWORK)
+      // Compose request body for each query
+      queryRequestBodyList.add(
+          query.request.operation.composeRequestBody(
+              query.request.autoPersistQueries,
+              query.request.sendQueryDocument,
+              scalarTypeAdapters
+          )
+      )
+    }
+    // Combine all request bodies into one
+    val batchRequestBody = RequestBody.create(
+        ApolloServerInterceptor.MEDIA_TYPE,
+        createBatchRequestJsonBody(queryRequestBodyList)
+    )
+    // TODO assumes all queries are POST (does GET even make sense for batch HTTP calls?)
```

Comment:
We never use GET for queries, so didn't implement it yet, but wondering what you think since batch request bodies can be quite large, feels wrong to put it all in the URL. Wondering if batching should only be enabled when using POST queries.

Reply:
Good question. Maybe throw an error if trying to enable `GET` + `batching` at the same time? My guess is that 99% of use cases will use `POST`, we can have the discussion when/if someone needs `GET` ?

Reply:
yeah that's what we were thinking 👍

Reply:
`apollo-client` web doesn't support Batch + GET either: https://github.com/apollographql/apollo-client/blob/main/src/link/batch-http/batchHttpLink.ts#L108